### PR TITLE
GH-35497: [C++] Use the latest tagged version of flatbuffers

### LINKED ISSUE #35497: [C++] Pin flatbuffers compiler version
### Describe the bug, including details regarding any error messages, version, and platform.

I ran cpp/build-support/update-flatbuffers.sh and the regenerated headers included new failing static assertions like:

```
/home/ben/arrow/cpp/src/generated/Message_generated.h:11:41: error: static assertion failed: Non-compatible flatbuffers version included
   11 | static_assert(FLATBUFFERS_VERSION_MAJOR == 23 &&
```

The probable fix is to pin the version of the flatbuffers compiler used in c++. As of https://github.com/google/flatbuffers/pull/7203 the compiler inserts these static assertions which assert the version of the compiler- but our vendored flatbuffers are at [1.12.0](https://github.com/apache/arrow/blob/a9a8626e8342e1145ec1b60ffb50154693b9ef92/cpp/thirdparty/flatbuffers/include/flatbuffers/base.h#L150-L152). conda-forge has that version and after installing it I don't see failing static assertions any more.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by felipecrv**
### Rationale for this change

To use a more modern version of `flatc` in Arrow.

### What changes are included in this PR?

1) Re-generating the the C++ files with a `flatc` based on the latest tag on the `flatbuffer` repo.
2) Copy the minimal set of includes to our vendored folder
3) Manually re-apply the patches that have been applied to the previous headers to fix issues

### Are these changes tested?

Tested by building everything correctly.
* Closes: #35497

### CODE REVIEW

File: cpp/thirdparty/flatbuffers/include/flatbuffers/allocator.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FLATBUFFERS_ALLOCATOR_H_
+#define FLATBUFFERS_ALLOCATOR_H_
+
+// Move this vendored copy of flatbuffers to a private namespace,
+// but continue to access it through the "flatbuffers" alias.
+namespace arrow_vendored_private {
```
Comment (pitrou) [inline cpp/thirdparty/flatbuffers/include/flatbuffers/allocator.h:22]: If we switch to C++17-style `arrow_vendored_private::flatbuffers` it probably reduces the patch size and may make automation more doable as well.

File: cpp/thirdparty/flatbuffers/README.md [hunk 2]
```diff
@@ -75,48 +117,229 @@ index fccce42f6..a00d5b0fd 100644
...
+-      #if !defined(ABSL_USES_STD_STRING_VIEW)
+-        #include "absl/strings/string_view.h"
+-        namespace flatbuffers {
+-          typedef absl::string_view string_view;
+-        }
+-        #define FLATBUFFERS_HAS_STRING_VIEW 1
+-      #endif
+-    #endif
+-  #endif // __has_include
++  #include <string_view>
```
Comment (pitrou) [inline cpp/thirdparty/flatbuffers/README.md:155]: Is this change actually necessary or would the original "Check for std::string_view (in c++17)" snippet above work for us? We probably want to minimize the patch size.

Comment (felipecrv) [reply cpp/thirdparty/flatbuffers/README.md:155]: You're right. Since this version doesn't check for `absl::string_view` first, we don't need it.

Comment (felipecrv) [reply cpp/thirdparty/flatbuffers/README.md:155]: Reference: https://github.com/apache/arrow/pull/12204

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp -g python

**Comment by pitrou** [comment]
CI failures look unrelated, I'll merge. Thanks @felipecrv !

**Comment by kou** [comment]
@github-actions crossbow submit test-skyhook-integration

**Comment by kou** [comment]
@felipecrv It seems that this broke `test-skyhook-integration`:

https://github.com/ursacomputing/crossbow/actions/runs/6502407264/job/17661381250#step:6:2369

```text
FAILED: src/skyhook/CMakeFiles/arrow_skyhook_objlib.dir/protocol/skyhook_protocol.cc.o 
/usr/local/bin/sccache /usr/bin/c++  -DARROW_HAVE_RUNTIME_AVX2 -DARROW_HAVE_RUNTIME_AVX512 -DARROW_HAVE_RUNTIME_BMI2 -DARROW_HAVE_RUNTIME_SSE4_2 -DARROW_HAVE_SSE4_2 -DARROW_MIMALLOC -DARROW_NO_DEPRECATED_API -DARROW_WITH_RE2 -DARROW_WITH_TIMING_TESTS -DARROW_WITH_UTF8PROC -DBOOST_ALL_NO_LIB -Isrc -I/arrow/cpp/src -I/arrow/cpp/src/generated -isystem /arrow/cpp/thirdparty/flatbuffers/include -isystem /arrow/cpp/thirdparty/hadoop/include -isystem google_cloud_cpp_ep-install/include -isystem absl_ep-install/include -isystem crc32c_ep-install/include -isystem orc_ep-install/include -isystem protobuf_ep-install/include -isystem awssdk_ep-install/include -isystem opentelemetry_ep-install/include -isystem xsimd_ep/src/xsimd_ep-install/include -isystem jemalloc_ep-prefix/src -isystem mimalloc_ep/src/mimalloc_ep/include/mimalloc-2.0 -Wno-noexcept-type  -fdiagnostics-color=always  -Wall -Wno-conversion -Wno-sign-conversion -Wunused-result -Wdate-time -fno-semantic-interposition -msse4.2  -g -Werror -O0 -ggdb -g1 -fPIC   -pthread -std=c++17 -MD -MT src/skyhook/CMakeFiles/arrow_skyhook_objlib.dir/protocol/skyhook_protocol.cc.o -MF src/skyhook/CMakeFiles/arrow_skyhook_objlib.dir/protocol/skyhook_protocol.cc.o.d -o src/skyhook/CMakeFiles/arrow_skyhook_objlib.dir/protocol/skyhook_protocol.cc.o -c /arrow/cpp/src/skyhook/protocol/skyhook_protocol.cc
In file included from /arrow/cpp/src/skyhook/protocol/skyhook_protocol.cc:21:
/arrow/cpp/src/skyhook/protocol/ScanRequest_generated.h: In member function 'bool org::apache::arrow::flatbuf::ScanRequest::Verify(arrow_vendored_private::flatbuffers::Verifier&) const':
/arrow/cpp/src/skyhook/protocol/ScanRequest_generated.h:45:55: error: no matching function for call to 'org::apache::arrow::flatbuf::ScanRequest::VerifyField<int64_t>(arrow_vendored_private::flatbuffers::Verifier&, org::apache::arrow::flatbuf::ScanRequest::FlatBuffersVTableOffset) const'
   45 |            VerifyField<int64_t>(verifier, VT_FILE_SIZE) &&
      |                                                       ^
In file included from /arrow/cpp/thirdparty/flatbuffers/include/flatbuffers/flatbuffer_builder.h:42,
                 from /arrow/cpp/thirdparty/flatbuffers/include/flatbuffers/flatbuffers.h:35,
                 from /arrow/cpp/src/skyhook/protocol/skyhook_protocol.cc:19:
/arrow/cpp/thirdparty/flatbuffers/include/flatbuffers/table.h:132:8: note: candidate: 'bool arrow_vendored_private::flatbuffers::Table::VerifyField(const arrow_vendored_private::flatbuffers::Verifier&, arrow_vendored_private::flatbuffers::voffset_t, size_t) const [with T = long int; arrow_vendored_private::flatbuffers::voffset_t = short unsigned int; size_t = long unsigned int]'
  132 |   bool VerifyField(const Verifier &verifier, voffset_t field,
      |        ^~~~~~~~~~~
/arrow/cpp/thirdparty/flatbuffers/include/flatbuffers/table.h:132:8: note:   candidate expects 3 arguments, 2 provided
In file included from /arrow/cpp/src/skyhook/protocol/skyhook_protocol.cc:21:
/arrow/cpp/src/skyhook/protocol/ScanRequest_generated.h:46:57: error: no matching function for call to 'org::apache::arrow::flatbuf::ScanRequest::VerifyField<int16_t>(arrow_vendored_private::flatbuffers::Verifier&, org::apache::arrow::flatbuf::ScanRequest::FlatBuffersVTableOffset) const'
   46 |            VerifyField<int16_t>(verifier, VT_FILE_FORMAT) &&
      |                                                         ^
In file included from /arrow/cpp/thirdparty/flatbuffers/include/flatbuffers/flatbuffer_builder.h:42,
                 from /arrow/cpp/thirdparty/flatbuffers/include/flatbuffers/flatbuffers.h:35,
                 from /arrow/cpp/src/skyhook/protocol/skyhook_protocol.cc:19:
/arrow/cpp/thirdparty/flatbuffers/include/flatbuffers/table.h:132:8: note: candidate: 'bool arrow_vendored_private::flatbuffers::Table::VerifyField(const arrow_vendored_private::flatbuffers::Verifier&, arrow_vendored_private::flatbuffers::voffset_t, size_t) const [with T = short int; arrow_vendored_private::flatbuffers::voffset_t = short unsigned int; size_t = long unsigned int]'
  132 |   bool VerifyField(const Verifier &verifier, voffset_t field,
      |        ^~~~~~~~~~~
/arrow/cpp/thirdparty/flatbuffers/include/flatbuffers/table.h:132:8: note:   candidate expects 3 arguments, 2 provided
```

Could you re-generate `cpp/src/skyhook/protocol/ScanRequest_generated.h` too?

**Comment by felipecrv** [comment]
> Could you re-generate cpp/src/skyhook/protocol/ScanRequest_generated.h too?

Sure. Let me create an issue and submit a PR soon.

**Pull request merged**